Move Speculation Rules to new location as purely a standalone plugin

felixarntz opened the pull request:
## Summary

See https://github.com/WordPress/performance/pull/904#issuecomment-1896444353: Since the Speculation Rules feature will not be published as a module (given the unbundling of Performance Lab) and thus only as a plugin, it should be moved to the new location for standalone plugins (see #934).

This PR is still against the feature branch and prepares that feature branch for a clean merge after #934 has been implemented. There's no blocker for merging this into the feature branch though.

**No functional changes are included as part of this PR.**

## Checklist

- [x] PR has either `[Focus]` or `Infrastructure` label.
- [x] PR has a `[Type]` label.
- [x] PR has a milestone or the `no milestone` label.

Commit:
Implement basic module to experiment with speculationrules for prerendering.

Commit:
Include proper backslashes in rules source.

Commit:
Add hook to make it easy to use origin trial.

Commit:
Merge branch 'trunk' into speculationrules-experiment

Commit:
Merge branch 'trunk' into speculationrules-experiment

Commit:
Print speculationrules script tag directly without using JS.

Commit:
Introduce filter for which paths to exclude for speculative prerendering.

Commit:
Add CODEOWNERS entry.

Commit:
Add initial test coverage for Speculation Rules.

Commit:
Fix function name caught as misspelling.

Commit:
Merge branch 'trunk' into speculationrules-experiment

Commit:
Address CR feedback.

Commit:
Add TODO regarding eventual wildcard removal.

Commit:
Add speculation rules URL pattern prefixer to support WordPress site in subdirectories.

Commit:
Add missing return type hint.

Commit:
Remove unnecessary relative_to directive.

felixarntz committed:
Simplify escaping.

Co-authored-by: Weston Ruter <westonruter@google.com>

felixarntz committed:
Merge pull request #733 from WordPress/speculationrules-experiment

Implement new module to use Speculation Rules API for prerendering documents on hover

felixarntz committed:
Merge pull request #936 from WordPress/trunk

Update feature branch with `trunk`

Commit:
Fix new WPCS violation after trunk refresh.

Commit:
Add readme.txt for Speculation Rules plugin.

Commit:
Add Speculation Rules settings and UI.

Commit:
Add references to UI for customization of Speculation Rules.

Commit:
Add link to WICG spec draft.

Commit:
Remove Speculation Rules origin trial logic.

Commit:
Remove now unnecessary path suffixes.

Commit:
Remove outdated suffixes from tests as well.

Commit:
Explicitly fall back to setting default if setting somehow missing (for defensive coding, mostly relevant for tests where setting is not registered).

Commit:
Only consider no-prerender class when mode is actually set to prerender.

Commit:
Remove irrelevant assertion.

felixarntz committed:
Add test coverage.

felixarntz committed:
Fix test.

mukeshpanchal27 committed:
Merge pull request #937 from WordPress/fix/speculation-rules-phpcs-violation

Fix new WPCS violation after trunk refresh

felixarntz committed:
Update grammar.

Co-authored-by: Jeremy Roman <jeremy@jeremyroman.com>

Commit:
Merge branch 'feature/speculationrules' into add/speculation-rules-readme

Commit:
Clarify behavior and browser support in readme.

Commit:
Merge branch 'feature/speculationrules' into add/speculation-rules-settings

Commit:
Clarify eager configuration.

Commit:
Clarify looseness of eagerness values.

Commit:
Clarify eagerness.

felixarntz committed:
Change link format.

Co-authored-by: Weston Ruter <westonruter@google.com>

felixarntz committed:
Merge pull request #938 from WordPress/add/speculation-rules-readme

Add `readme.txt` for Speculation Rules plugin

felixarntz committed:
Merge pull request #940 from WordPress/enhance/speculation-rules-chrome-121

Removes outdated pieces of Speculation Rules logic that are no longer needed as of Chrome 121

felixarntz committed:
Merge branch 'feature/speculationrules' into add/speculation-rules-settings

felixarntz committed:
Use more precise array doc type annotation.

Co-authored-by: Weston Ruter <westonruter@google.com>

Commit:
Use more precise array doc syntax in more places.

Commit:
Limit maximum width instead of using br.

Commit:
Use better phrased description for REST API too.

Commit:
More defensive coding when accessing array keys.

felixarntz committed:
Merge pull request #939 from WordPress/add/speculation-rules-settings

Add Speculation Rules settings and UI

Commit:
Move Speculation Rules to new location as purely a standalone plugin.

swissspidy reviewed (approved):
I take it the lint errors are known?

felixarntz commented:
I realize there's a PHPStan error here, which may be addressed by including the `plugins/` directory in `phpstan.neon.dist`, however it pointed me to another problem: We should probably move the test coverage to another place as well.

This also relates to unit testing. We'll have to make a few changes to the infrastructure, either change `bootstrap.php` somehow or use different PHPUnit config files based on which plugin is tested. There's probably some way to do this with the least code duplication possible, maybe using environment variables or something. I think it makes sense to discuss that further as part of #934.

jeremyroman committed:
Escape path prefix and restrict it to be a pathname in Speculation Rules

On unusual deployment setups (but notably WordPress Playground), the
path prefix might contain one of the URL pattern special characters,
such as :.

Escape these characters in the path prefix, and since some of these also
confuse the shorthand constructor string syntax, use the longhand syntax
which specifies that the presence of a : here does not indicate a
protocol/scheme.

westonruter committed:
Force HTML5 script theme support when printing JSON script

westonruter committed:
Negate condition

westonruter committed:
Add tests for valid JSON workaround

westonruter committed:
Add ticket reference

westonruter committed:
Limit workaround to WP 6.4

jeremyroman committed:
Use braces to deal with component confusion issue instead (and other code review comments).

jeremyroman committed:
Use addcslashes

Co-authored-by: Weston Ruter <westonruter@google.com>

adamsilverstein committed:
Update modules/js-and-css/speculation-rules/class-plsr-url-pattern-prefixer.php

Co-authored-by: Weston Ruter <westonruter@google.com>

westonruter committed:
Merge pull request #951 from jeremyroman/sr-escape-pathname

Escape path prefix and restrict it to be a pathname in Speculation Rules

westonruter committed:
Guard against alpha, beta, and RC from interfering with versions

westonruter committed:
Ensure theme features are cleaned up after tests

westonruter committed:
Merge pull request #952 from WordPress/fix/non-html5-script-printing

Force HTML5 script theme support when printing JSON script

westonruter committed:
Add plugin directory assets for Speculation Rules

westonruter committed:
Squoosh images from Figma

westonruter committed:
Compress icon.svg with SVGOMG

swissspidy committed:
Merge pull request #987 from WordPress/add/speculation-rules-plugin-assets

Add plugin directory assets for Speculation Rules

westonruter committed:
Bump version and update readme in Speculation Rules

westonruter committed:
Merge branch 'feature/speculationrules' of https://github.com/WordPress/performance into prepare/speculation-rules-standalone-plugin

westonruter commented on plugins/speculation-rules/load.php:
```diff
@@ -0,0 +1,28 @@
+ * Description: Uses the Speculation Rules API to prerender linked URLs upon hover by default.
+ * Requires at least: 6.3
+ * Requires PHP: 7.0
+ * Version: 1.0.0
```
After https://github.com/WordPress/performance/pull/993 this will need to be:

```suggestion
 * Version: 1.0.1
```

westonruter commented on plugins/speculation-rules/load.php:
```diff
@@ -0,0 +1,28 @@
+	return;
+}
+
+define( 'SPECULATION_RULES_VERSION', '1.0.0' );
```
Ditto:

```suggestion
define( 'SPECULATION_RULES_VERSION', '1.0.1' );
```

westonruter committed:
Fix typo in changelog

Co-authored-by: Pascal Birchler <pascalb@google.com>

westonruter committed:
Merge pull request #993 from WordPress/update/speculation-rules-changelog

Bump version and update readme in Speculation Rules

Commit:
Merge branch 'feature/speculationrules' into prepare/speculation-rules-standalone-plugin

Commit:
Update hard-coded Speculation Rules version to 1.0.1.

felixarntz commented:
@westonruter @swissspidy @mukeshpanchal27 Since this is the only remaining PR at the moment for Speculation Rules and we're working on the standalone plugins infrastructure in https://github.com/WordPress/performance/tree/feature/modules-to-plugins (which already includes the `auto-sizes` plugin), I will change the base branch of this PR to be that one.

The `feature/speculationrules` branch will then be obsolete and can be deleted, all further development of the Speculation Rules plugin can happen against `feature/modules-to-plugins` and, once that is merged, against `trunk`.

Commit:
Merge branch 'feature/modules-to-plugins' into prepare/speculation-rules-standalone-plugin

Commit:
Move Speculation Rules to tests/plugins and refresh setup based on latest infrastructure.

felixarntz commented on .github/workflows/php-test-plugins.yml:
```diff
@@ -76,3 +76,7 @@ jobs:
         run: npm run test-php -- -- -- --testsuite auto-sizes
       - name: Running multisite unit tests for Auto-sizes for Lazy-Loaded Images Plugin
         run: npm run test-php-multisite -- -- -- --testsuite auto-sizes
+      - name: Running single site unit tests for Speculation Rules Plugin
+        run: npm run test-php -- -- -- --testsuite speculation-rules
+      - name: Running multisite unit tests for Speculation Rules Plugin
+        run: npm run test-php-multisite -- -- -- --testsuite speculation-rules
```
This is cumbersome so we should probably add shortcuts via scripts in `composer.json` and/or `package.json`.

There should also be a shortcut that runs the single site and multisite tests respectively for _all_ of these plugins at once, so that we don't have to add every single plugin to this GitHub workflow file.

Let's handle that in a separate PR though, not directly related to the Speculation Rules work.

felixarntz replied:
I opened #1017 for that.

felixarntz commented on phpcs.xml.dist:
```diff
@@ -8,7 +8,7 @@
 	<rule ref="WordPress-Extra" />
 	<rule ref="WordPress.WP.I18n"/>
-	<config name="text_domain" value="performance-lab,default,auto-sizes"/>
+	<config name="text_domain" value="performance-lab,default,auto-sizes,speculation-rules"/>
```
This will need to be adjusted once #1002 is merged, instead a `plugins/speculation-rules/phpcs.xml.dist` file should be added then.

felixarntz reviewed (commented):
@westonruter @swissspidy @mukeshpanchal27 This is pretty much ready now and refreshed for the way things work now in `feature/modules-to-plugins`.

We should probably merge #1002 first, then do another quick refresh of this one, then merge this one.

After that we'll have both `auto-sizes` and `speculation-rules` in the `feature/modules-to-plugins` branch and can continue development against that branch until it's merged into `trunk`.

westonruter commented:
> We should probably merge #1002 first, then do another quick refresh of this one, then merge this one.

Let's merge it then. 🚀

westonruter commented:
> Let's merge it then. 🚀

Merged!

Commit:
Merge branch 'feature/modules-to-plugins' into prepare/speculation-rules-standalone-plugin

Commit:
Update composer.json shortcuts to include speculation-rules.

Commit:
Include phpcs.xml.dist for speculation-rules.

Commit:
Update composer.json shortcuts to include speculation-rules linting.

felixarntz commented:
@westonruter @swissspidy @mukeshpanchal27 Should be ready for merge now!

You had already approved prior, but since I now changed the base branch, it would be great if you could do another quick sanity check.

Note that all the actual PHP code for Speculation Rules is unchanged and was already approved in the `feature/speculationrules` branch, so you can ignore the PHP code for the review.

felixarntz commented:
@westonruter It looks like a test related to the CDATA fix you implemented fails, see https://github.com/WordPress/performance/actions/runs/8071362067/job/22050829990?pr=946. Maybe it is because the test code isn't limited to the specific WordPress versions?

Since the CDATA issue is only a problem on WordPress 6.4, we probably need to limit the test accordingly. Either we make those assertions conditional, or we mark the entire test to skip unless it's WordPress version 6.4.x. WDYT?

On another note: I think the code comment linking to the Trac ticket points to the wrong ticket? Shouldn't it be https://core.trac.wordpress.org/ticket/60320?

westonruter committed:
Add missing ext-json requirement to composer.json

westonruter committed:
Run composer update after adding ext-json

westonruter committed:
Fix ticket reference and add covers

westonruter committed:
Fix test for CDATA

westonruter commented on tests/plugins/speculation-rules/plsr-url-pattern-prefixer-test.php:
```diff
@@ -0,0 +1,76 @@
+/**
+ * Tests for PLSR_URL_Pattern_Prefixer class.
+ *
+ * @package performance-lab
+ * @group speculation-rules
```
```suggestion
 * @package speculation-rules
```

westonruter commented on tests/plugins/speculation-rules/speculation-rules-helper-test.php:
```diff
@@ -0,0 +1,93 @@
+/**
+ * Tests for speculation-rules helper file.
+ *
+ * @package performance-lab
+ * @group speculation-rules
```
```suggestion
 * @package speculation-rules
```

westonruter commented on tests/plugins/speculation-rules/speculation-rules-settings-test.php:
```diff
@@ -0,0 +1,96 @@
+/**
+ * Tests for speculation-rules settings file.
+ *
+ * @package performance-lab
+ * @group speculation-rules
```
```suggestion
 * @package speculation-rules
```

westonruter commented on tests/plugins/speculation-rules/speculation-rules-test.php:
```diff
@@ -0,0 +1,60 @@
+/**
+ * Tests for speculation-rules module.
+ *
+ * @package performance-lab
+ * @group speculation-rules
```
```suggestion
 * @package speculation-rules
```

westonruter commented:
@felixarntz I opted to keep the test but add a condition for the WP version. Tests are passing now.

westonruter reviewed (approved):
Just some trivial phpdoc changes for the test files, but otherwise I'm pre-approving.

felixarntz committed:
Update package name in speculation-rules tests.

Co-authored-by: Weston Ruter <westonruter@google.com>

mukeshpanchal27 commented on plugins/speculation-rules/class-plsr-url-pattern-prefixer.php:
```diff
@@ -0,0 +1,122 @@
+ * Class 'PLSR_URL_Pattern_Prefixer'.
+ *
+ * @package speculation-rules
+ * @since n.e.x.t
```
What about version? When we change it from `n.e.x.t` to actual version `1.0.0`

swissspidy replied:
Since the plugin is already published, this should be already 1.0.0, no?

mukeshpanchal27 commented on plugins/speculation-rules/load.php:
```diff
@@ -0,0 +1,28 @@
+<?php
+/**
+ * Plugin Name: Speculation Rules
+ * Plugin URI: https://github.com/WordPress/performance/tree/trunk/modules/js-and-css/speculation-rules
```
```suggestion
 * Plugin URI: https://github.com/WordPress/performance/tree/trunk/plugins/speculation-rules
```

felixarntz replied:
Good catch! I fixed this for `auto-sizes` as well.

Commit:
Fix incorrect Plugin URI for new standalone plugins.

Commit:
Replace n.e.x.t in Speculation Rules with the actual version numbers.

felixarntz commented:
@swissspidy @mukeshpanchal27 I've addressed your feedback! This is probably good to merge now.
